### PR TITLE
Implement AntiCheat System - AntiHacks

### PR DESCRIPTION
**Pull request by kvipka**
**Updated version with log in auth database for fresh TC revision**

-ASH - AntiSpeedHack (+anti teleport hack)
-AFH - AntiFlyHack
-DoubleJump
-Wallclimb
-FakeJumper
-FakeFlyingMode

**Updated by 07 sept 2019:**
- implement ASH control for vehicle and transport/elevators
- implement 2 new movementgenerators : 
* JumpMovementGenerator
* ChargeMovemengGenerator
it will give correct information to server of unit-status and correct unit state. Also it fixed some issues (look info in comments)

This system has been tested on some big projects with big online (project names are hidden). 
**Issues:** No false positives

For some questions : discord - 
https://discord.gg/GcJJfXG

youtube videos : https://www.youtube.com/watch?v=aZP3qm9Ql0Y&list=PLUpIDd4JDWyOOJhJRcfaVL0RaRH1Ra5Xy

Note:
This system using only mathematic operations, without additional manager as AC2, no using additional memory.

Part of code can be shorter, when [PR](https://github.com/TrinityCore/TrinityCore/pull/18771) with ACK-packet handlers will complete.

ACK 1

Log in auth-database :
Безымянный

### CODE REVIEW

**Comment by illfated** [comment]
Ref. #23509 (previous PR with the same content)

**Comment by Jildor** [comment]
I'm not sure if log detection in the auth db is a god idea (performance things).
But maybe it's better if you add a check for enable/disable this option by the conf file ;)

**Comment by kvipka** [comment]
> performance things

log in database by AccountMgr (another thread) + by encapsulated asynchronous query. I think no reason to worry

**Comment by jackpoz** [comment]
how come you create a new PR for the same code ? it makes the review and discussion about the code confusing

**Comment by illfated** [comment]
Oh. Is the content here without the changes in the previous PR (which was closed due to file conflicts)?

**Comment by kvipka** [comment]
> Oh. Is the content here without the changes in the previous PR (which was closed due to file conflicts)?

"with"

**Comment by illfated** [comment]
OK, fair enough. I misinterpreted the comment jackpoz wrote.

**Comment by jackpoz** [comment]
> **Issues:** No false positives

does this new PR fix all the false positives reported in the old PR https://github.com/TrinityCore/TrinityCore/pull/23509 ?

**Comment by kvipka** [comment]
From previous pr, yes

**Comment by Infernales** [comment]
@kvipka I don’t remember if I asked such a question: should anticheat determine the hack of walking on water and the hack of jump height?

**Comment by kvipka** [comment]
> @kvipka I don’t remember if I asked such a question: should anticheat determine the hack of walking on water and the hack of jump height?

nope and nope

**Comment by Infernales** [comment]
@kvipka Well then, everything is in order.

**Comment by Infernales** [comment]
@kvipka False positives:
```
Unit::CheckMovementInfo :  Speed hack detected for Account Id : 1, Player Test
Unit::CheckMovementInfo :  oldX = 16222.885742
Unit::CheckMovementInfo :  oldY = 16260.132812
Unit::CheckMovementInfo :  newX = 16222.971680
Unit::CheckMovementInfo :  newY = 16259.600586
Unit::CheckMovementInfo :  packet distance = 0.539120
Unit::CheckMovementInfo :  available distance = 0.483302
Unit::CheckMovementInfo :  move time = 1249580288.000000
Unit::CheckMovementInfo :  delay sent ptk - recieve pkt (previous) = 0.000000
Unit::CheckMovementInfo :  full delay = 0.000038
Unit::CheckMovementInfo :  diff time = 0.060038
Unit::CheckMovementInfo :  real ping = 0
Unit::CheckMovementInfo :  fake ping = 60

Unit::CheckMovementInfo :  Speed hack detected for Account Id : 1, Player Test
Unit::CheckMovementInfo :  oldX = 16227.772461
Unit::CheckMovementInfo :  oldY = 16255.753906
Unit::CheckMovementInfo :  newX = 16227.333008
Unit::CheckMovementInfo :  newY = 16256.427734
Unit::CheckMovementInfo :  packet distance = 0.804465
Unit::CheckMovementInfo :  available distance = 0.483302
Unit::CheckMovementInfo :  move time = 1249650048.000000
Unit::CheckMovementInfo :  delay sent ptk - recieve pkt (previous) = 0.000000
Unit::CheckMovementInfo :  full delay = 0.000038
Unit::CheckMovementInfo :  diff time = 0.060038
Unit::CheckMovementInfo :  real ping = 0
Unit::CheckMovementInfo :  fake ping = 60
```
Anticheat falsely triggered when the client started on the same machine as the server. Well, a classic of the genre: false positives began after 14 days of the uptime server.

**Comment by jackpoz** [comment]
does that happen when the OS updates the clock ?

**Comment by Infernales** [comment]
> does that happen when the OS updates the clock ?
No. This happened when starting the torrent client. When I tried to connect to the server from a laptop, the same situation arose there too, but already because of the Windows update.

**Comment by illfated** [comment]
Sounds like you created your own local lag situation. Could be useful for testing, provoking lag, but not sure how much it tells us about the PR content.

**Comment by Infernales** [comment]
@illfated So the problem is that even after closing the torrent client, it continues to kick from the server.

**Comment by illfated** [comment]
I understand that I should avoid guessing what is going on with the code, but could this have anything to do with the size of your logs or the number of logs being written by the servers? (Comparing to how Blizzard servers restart at least once every week, although they usually have a lot of players online.)

**Comment by Infernales** [comment]
@illfated It is unlikely, but yes, after restarting the server(TrinityCore), everything works again.

**Comment by illfated** [comment]
Fair enough, thank you for confirming that. At least we know it is something that builds up over time (and usually more than a week or 10 days).

**Comment by kvipka** [comment]
WoWScrnShot_090519_172704

Uptime 15 days, no troubles

**Comment by Infernales** [comment]
@kvipka Strange, maybe the problem is that I compiled the server with the -DWITH_DYNAMIC_LINKING=1 and -DSCRIPTS="dynamic" parameters? Yes, I forgot to say about it.

**Comment by Jildor** [comment]
👍 
Can anyone test vehicle speed with auras tha increase vehicle speed movement like Ulduar - Salvaged Chopper (33062) with 62299 aura?

File: src/server/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -4065,3 +4066,116 @@ Metric.OverallStatusInterval = 1
...
+###################################################################################################
+#
+# AntiCheats.forceExcludeMapsid disabled maps
+#        Description: disabled mapid for anticheat system
+#        Example:     "616,649"
+
+AntiCheats.forceExcludeMapsid = ""
+
+#
+###################################################################################################
```
Comment (Aokromes) [inline src/server/worldserver/worldserver.conf.dist]: missing new line :)

Comment (kvipka) [reply src/server/worldserver/worldserver.conf.dist]: > missing new line :)

tenor

File: sql/custom/auth/2019_07_24_00_anticheat_logs.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+CREATE TABLE `anticheat_logs`  (
+  `id` int(10) UNSIGNED NOT NULL AUTO_INCREMENT,
+  `times` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  `account` int(9) NOT NULL,
+  `player` varchar(96) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  `description` varchar(100) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  `position` varchar(255) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL,
+  `realmId` int(9) NOT NULL,
+  PRIMARY KEY (`id`)
+) ENGINE = InnoDB AUTO_INCREMENT = 0 CHARACTER SET = utf8;
```
Comment (illfated) [inline sql/custom/auth/2019_07_24_00_anticheat_logs.sql]: Missing NewLine ("No newline at end of file").

**Comment by kvipka** [comment]
Implement 2 new movementgenerators : JumpMovementGenerator and ChargeMovementGenerator

Fix false/positives for charge/deathgrip and others jump spells (KnockbackFrom and leaps)

Now Charge and Jump are not virtual movement. And server will know, when movement will end.
Also fixed some issues : 
- troubles with SAI when dk  grip this NPC
- troubles with NPC, when they drops under textures (for example knockback spells like shaman spell)

**Comment by Infernales** [comment]
@kvipka Build error:
```
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Movement/MovementGenerators/FollowMovementGenerator.cpp:204:55: fatal error: 

      no member named 'HasWalkingFlag' in 'Player'

                walk = owner->GetOwner()->ToPlayer()->HasWalkingFlag();

                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^

1 error generated.

src/server/game/CMakeFiles/game.dir/build.make:2527: recipe for target 'src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/FollowMovementGenerator.cpp.o' failed

make[2]: *** [src/server/game/CMakeFiles/game.dir/Movement/MovementGenerators/FollowMovementGenerator.cpp.o] Error 1
```

**Comment by kvipka** [comment]
Implement WalkingFlag. What is it : 
* When player with Pet are walking - Pet walking too. When Player starting to run - pet starting to run too with owner. In TC it another issue. 

Just too lazy to look for all these issues. =)

**Comment by kvipka** [comment]
very lazy to write a description of last commit. Just look this videos : 

Async position Before : 
https://youtu.be/WJe24fYTgkw?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw
After : 
https://youtu.be/7dmc9-Nsrys?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw

Fear Before : 
https://youtu.be/vCc8SVWgWsk?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw
Fear after : 
https://youtu.be/Q9T_tnInBl0?list=PLUpIDd4JDWyPuzg5PyVuXoHx6sn7eqgYw

Just present...

**Comment by kvipka** [comment]
some relations : 
https://github.com/TrinityCore/TrinityCore/issues/21255

**Comment by Jildor** [comment]
wait, you're fixing a issues related with movement in PR of anticheat implementation? ( 👍 ❤️ )
isn't better split in a separated PR?

**Comment by illfated** [comment]
Fair point, because if something in the "fix movement" part needs reverting, it would be much easier to check by removing a specific commit from the main branch.
If you need the fixes to be present for this AntiCheat PR to work, ask for the movement improvement PR to be merged first.

**Comment by Infernales** [comment]
@illfated I think he did it because he has to tear out his code in pieces from his fork. In this case, it makes sense to do everything in one PR. Again, sooner or later this PR will be in branch 3.3.5. At least I really hope so.

**Comment by kvipka** [comment]
i will not separate charge/jump movegens.
Current TC has virtual realization - and projects has issues with grip/charge - because server don't know real position of player.

Early (and in PR for azerothcore) it handled by HACK : just timer of ACK packets, when ASH are skipping checking in this time.

This hack do possible to use teleporthacks. In this PR i removed this hack, and will not use something like this. 
Only full correct realization.
____

About Confuse/Fleeing improvement - i will not separate it too, it present.
p.s. If you don't want to use it - just don't cherry-pick last commit. As i know - this PR will never merged by TC. And by this reason all ppl can take code, what want from this PR.

**Comment by illfated** [comment]
Fair point too. Thank you for clarifying.

**Comment by kvipka** [comment]
Screenshot_20190908-121642



Gendalf and his magic

**Comment by Infernales** [comment]
> Uptime, no trouble with time. Server Ubuntu
Try configuring Cmake with the -DWITH_DYNAMIC_LINKING=1 and -DSCRIPTS="dynamic" options.
Perhaps because of this I had problems with anticheat when I connected to the server from the same PC.

**Comment by Aokromes** [comment]
i got reports of gundrak rinos doing "invisible" charges with the latest version of anticheat.

**Comment by kvipka** [comment]
> i got reports of gundrak rinos doing "invisible" charges with the latest version of anticheat.

what the rinos? more info need)

**Comment by illfated** [comment]
- https://www.wowhead.com/npc=29931/drakkari-rhino
- https://www.wowhead.com/npc=29838/drakkari-rhino

**Comment by Aokromes** [comment]
https://mega.nz/#!x2pBGKLK!5QVu47oYg1jKq_sFTfXMDs6htBw1oNyMDeXp7uKS4kM

File: src/server/worldserver/worldserver.conf.dist [hunk 3]
```diff
@@ -4119,8 +4119,8 @@ AntiCheats.IgnoreControlMovement.Enabled = 1
 ###################################################################################################
 # Logs for Anticheat
 
-Appender.Anticheat=2,3,15,Anticheat.log
-Logger.anticheat=3,Console Anticheat
+#Appender.Anticheat=2,3,15,Anticheat.log
```
Comment (kvipka) [inline src/server/worldserver/worldserver.conf.dist]: Now config with additional file with logs commited. All info in auth-database in anticheat_logs.

Comment (jackpoz) [reply src/server/worldserver/worldserver.conf.dist]: are old logs ever cleared ?

Comment (kvipka) [reply src/server/worldserver/worldserver.conf.dist]: > are old logs ever cleared ?

what you mean? .log files? as you wish. If about data in anticheat_logs - nope.

**Comment by wowmane** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/2088#issuecomment-531731970

https://github.com/azerothcore/azerothcore-wotlk/pull/2088#issuecomment-529208280

**Comment by Cobus420** [comment]
are there any new failures/bugs here or is this pull ready?

**Comment by illfated** [comment]
@Cobus420 : Feedback from individual testers is still useful at this stage, even if this PR is marked as ready ( **[Feedback-PatchFix]** ).

**Comment by Infernales** [comment]
@kvipka After 15 days, everything kicks the same way, if you go to the server from the same PC where the server itself is running. 
@anyone Can anyone else confirm this except me?

**Comment by kvipka** [comment]
> @kvipka After 15 days, everything kicks the same way, if you go to the server from the same PC where the server itself is running.
> Can anyone else confirm this except me?

can you give answer on this questions :
- do you use windows ?
- if yes, windows has auto correction of time?

**Comment by Infernales** [comment]
@kvipka I used three different servers with Arch Linux, Debian and CentOS.

**Comment by Jildor** [comment]
When you're play in the same machine that server, you have a very low ms, maybe can be this the cause?

**Comment by Jildor** [comment]
oh after 15 days,I did not noticed it, I don't know then xD

**Comment by Jinnaii** [comment]
Is this 15 days situation the only thing that stops this PR from merging or or are there a few unresolved false positive left?
In my opinion a Server running for 15+ days is not a realistic example but i may be wrong.

**Comment by Cobus420** [comment]
I got the Server Running for 18 Days with that PR. I dont get any issues with kicks or errors. Running on Windows Server.

**Comment by Aokromes** [comment]
it false positives a lot with people with >300ms lag

**Comment by qyh214** [comment]
like @Aokromes  said, lag is 500-700ms,I three jumps(space x 3) then I kick and disconnect with server,

Temporary fix: kick disabled

**Comment by Jinnaix** [comment]
Even if this would mean to disable autokick duo to latency false positives, this system is a major improvement to what trinity offers now. Its worth to have for logging and manual banning afterwards.

Is there any way to get and log player latency?

**Comment by Jildor** [comment]
yep 
```
GetSession()->GetLatency();
```

**Comment by kvipka** [comment]
okay, rewrited algorithm, now should be okay

**Comment by Aokromes** [comment]
cannot be merged after latest changes.

Conflicting files
src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp

**Comment by jackpoz** [comment]
It seems this pr reverts some old MotionMaster.cpp changes too

**Comment by kvipka** [comment]
okay, rdy

**Comment by jackpoz** [comment]
I would say JumpMovementGenerator and ChargeMovemengGenerator should be in their own PR as they have little to do with anti-cheat

**Comment by kvipka** [comment]
Nope, it needed for anticheat

**Comment by jackpoz** [comment]
you don't need to remove it, more like open a PR with just those bits so it can be merged before this one

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Spells/Spell.cpp

**Comment by Aokromes** [comment]
lol it detected my own player character having 35ms lag.

**Comment by jackpoz** [comment]
> 3.500280, distance from packet = 3.50172

maybe it should be less strict than 0.00144 yards

File: src/server/game/Entities/Player/Anticheat.cpp [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+        delay = ptime - oldctime;
+        diffPacketdelay = 10000000 - delay;
+
+        if (oldctime > ptime)
+        {
+            TC_LOG_INFO("anticheat", "oldctime > ptime");
+            delay = 0;
+        }
+        diffPacketdelay = diffPacketdelay * 0.0000000001f;
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Anticheat.cpp]: what is the deal with all these 10000000 and 0.0000000001f numbers ? what are they supposed to do ?

**Comment by kvipka** [comment]
Безымянный

11 days of uptime, has no troubles yet

**Comment by illfated** [comment]
in case of preparation for merging, I presume there is a missing pch dependency:
```cpp
/home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
 fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
            if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
                                                       ^
1 error generated.

**Comment by kvipka** [comment]
> in case of preparation for merging, I presume there is a missing pch dependency:
> 
> ```c++
> /home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
>  fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
>             if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
>                                                        ^
> 1 error generated.
> ```
maybe because this ?
https://github.com/TrinityCore/TrinityCore/commit/be65f7d60e66d4c43f8fe9d779e5b4214afb91bf

smartscripts have not related with anticheats =)

**Comment by jackpoz** [comment]
>11 days of uptime, has no troubles yet

Did it report any real cheater or any false positive ?

**Comment by illfated** [comment]
> > in case of preparation for merging, I presume there is a missing pch dependency:
> > ```c++
> > /home/circleci/project/src/server/game/AI/SmartScripts/SmartScriptMgr.cpp:1508:56:
> >  fatal error: use of undeclared identifier 'MAX_MOVE_TYPE'
> >             if (e.action.movementSpeed.movementType >= MAX_MOVE_TYPE)
> >                                                        ^
> > 1 error generated.
> > ```
> 
> maybe because this ?
> [be65f7d](https://github.com/TrinityCore/TrinityCore/commit/be65f7d60e66d4c43f8fe9d779e5b4214afb91bf)
> 
> smartscripts have not related with anticheats =)

I know, but **ci/circleci: nopch** is currently reacting to the file `src/server/game/AI/SmartScripts/SmartScriptMgr.cpp` inside your PR branch.
I guess it doesn't really matter that much before the final merge takes place.

**Comment by kvipka** [comment]
https://www.youtube.com/watch?v=VYy-_pKrTnY

here you can see player with very bad inet connection and big ping interval ("ping horse racing")

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Spells/Spell.cpp

**Comment by Aokromes** [comment]
btw

git apply ../p3/anticheat
../p3/anticheat:213: trailing whitespace.
// Anticheat System
../p3/anticheat:269: trailing whitespace.
    if (GetAreaId() == 4859) // Area: 4859 (Frozen Throne)
../p3/anticheat:419: trailing whitespace.

../p3/anticheat:1393: trailing whitespace.
        Add(new JumpMovementGenerator<Player>(id, x, y, z, o, speedXY, max_height, hasOrientation));
../p3/anticheat:1970: trailing whitespace.

warning: squelched 7 whitespace errors
warning: 12 lines add whitespace errors.

**Comment by Aokromes** [comment]
still to picky
SpeedHack Detected for xxxxxx, normal distance for this time and speed = 3.465000, distance from packet = 3.465672

**Comment by Infernales** [comment]
Jumped in the elevator in Mulgore:
```
Player::CheckMovementInfo :  Wall climb hack detected for account Id : 1, Player Test, diffZ = 26.858719, distance = 0.000000, angle = 0.000000, Map = Kalimdor, mapId = 1, X = -5399.479492, Y = -2506.116699, Z = -39.996765
```
When using `.gm fly on` and mount, if you land on the ground, you get this:
```
2020-01-12_16:50:19 Unit::CheckMovementInfo :  Speed hack detected for account Id : 1, Player Test
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldX = 16228.923828
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldY = 16254.991211
2020-01-12_16:50:19 Unit::CheckMovementInfo :  newX = 16226.244141
2020-01-12_16:50:19 Unit::CheckMovementInfo :  newY = 16256.898438
2020-01-12_16:50:19 Unit::CheckMovementInfo :  packet distance = 3.289109
2020-01-12_16:50:19 Unit::CheckMovementInfo :  available distance = 1.652000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldStime = 401958430
2020-01-12_16:50:19 Unit::CheckMovementInfo :  oldCtime = 402069177
2020-01-12_16:50:19 Unit::CheckMovementInfo :  server time = 401958630
2020-01-12_16:50:19 Unit::CheckMovementInfo :  packet time = 402069412
2020-01-12_16:50:19 Unit::CheckMovementInfo :  diff delay between old ptk and current pkt = 0.001000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  full delay = 0.235000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  diff time = 0.236000
2020-01-12_16:50:19 Unit::CheckMovementInfo :  ping = 9
```

**Comment by Aokromes** [comment]
still:
 git apply ../p3/anticheat
../p3/anticheat:275: trailing whitespace.
    if (GetAreaId() == 4859) // Area: 4859 (Frozen Throne)
../p3/anticheat:1660: trailing whitespace.
        Add(new JumpMovementGenerator<Player>(id, x, y, z, o, speedXY, max_height, hasOrientation));
../p3/anticheat:2246: trailing whitespace.

../p3/anticheat:2260: trailing whitespace.

../p3/anticheat:2345: trailing whitespace.

warning: squelched 4 whitespace errors
warning: 9 lines add whitespace errors.

**Comment by Aokromes** [comment]
lately i am getting reports of people taking damage on land from flying mounts.

**Comment by Aokromes** [comment]
https://gist.github.com/Aokromes/3488abafe772e5b10a480cd9ba54c209

plrMover->SetFallInformation(plrMover->GetPositionZ());

is crashing core acording shauren.

**Comment by kvipka** [comment]
good catch!

**Comment by Aokromes** [comment]
i am still getting damage reports on landing.

**Comment by kvipka** [comment]
> i am still getting damage reports on landing.

pls more info, how to reproduce in future,maybe screens

**Comment by Aokromes** [comment]
https://mega.nz/#!xmZjiSYA!u0t8RjWyQN6lw2N1OUzWpI8-bGOE2ImVEX9jwoGYsLE

**Comment by kvipka** [comment]
Good, what addon of frame if nit a secret? :)

**Comment by Aokromes** [comment]
this false positives a lot.

MovementHandler::NoFallingDamage by Account id : , Player 
MovementHandler::NoFallingDamage by Account id : , Player 
MovementHandler::NoFallingDamage by Account id : , Player

**Comment by kvipka** [comment]
It interesting. This announce will working only when player, which have flag of falling_far, have lost this flag without sending land or swim opcode and without calling function ResetFallingInfo.In this time server start timer, while will waiting this opcode from player. When timer expired, server will send this announce and kick player. How player can receive this announce without hack idk. Can you ask him to do screenshots or better to do video?

**Comment by Aokromes** [comment]
i asked the guy bug he doesnt' remember exact place (i asked him 3h after got reported)

**Comment by Aokromes** [comment]
i still get a lot of false positives of 
MovementHandler::NoFallingDamage by Account id : , Player
and of players with high latency.

**Comment by Aokromes** [comment]
meh https://i.imgur.com/azHF9n0.jpg

**Comment by kvipka** [comment]
fixed issue when dk trying to do deathgrip on npc, but unit starting cast (for example scarlet priest).

**Comment by jackpoz** [comment]
This PR has a lot of changes all around the core and it might never be actually merged. What will most likely happen is:
1) The PR is left open here, waiting and waiting
2) The PR is pushed at https://github.com/TrinityCore/TrinityCoreCustomChanges as a branch that anyone can choose to use , listed at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki

Do you have any preference out of these 2 choices @kvipka ?

**Comment by kvipka** [comment]
@jackpoz i tihnk 1 variant are better, it's not a custom feature, just security.

**Comment by Aokromes** [comment]
i think this breaks flame leviathan fight, it doesn't even agro after breaking door.

**Comment by kvipka** [comment]
> i think this breaks flame leviathan fight, it doesn't even agro after breaking door.

how movement check can be related with bug with aggro-system ?

**Comment by jackpoz** [comment]
Like https://github.com/TrinityCore/TrinityCore/commit/3d11de239c294259bc9b6ec5c2638050fea0bf65 maybe ?

**Comment by Aokromes** [comment]
> > i think this breaks flame leviathan fight, it doesn't even agro after breaking door.
> 
> how movement check can be related with bug with aggro-system ?

well, remind invisible charges on gundrak xd this is the only big patch i have on my server non-present on tc test server.

**Comment by Aokromes** [comment]
> > i think this breaks flame leviathan fight, it doesn't even agro after breaking door.
> 
> how movement check can be related with bug with aggro-system ?

the only custom patch on this test server.

https://mega.nz/#!Mm5VyYYK!dvCNHg0W14pGo336OnhXRXtfl2xdSJfGZkhEwmSa2d4

**Comment by Aokromes** [comment]
[11:35:24] AntiCheat: Wallclimb Detected for Account id : , Player , diffZ = 22.276047, distance = 0.000000, angle = 0.000000, Map = Gundrak, mapId = 604, X = 1834.008911, Y = 858.692688, Z = 127.811676
i was told he dispelled priest levitate while falling.

**Comment by Rushor** [comment]
@kvipka do you have anything against PQR ?

**Comment by Aokromes** [comment]
a warrior charging while being stunned on the road on gundrak:
[21:46:01] AntiCheat: NoFallingDamage Hack Detected for Account : , Player :

**Comment by Aokromes** [comment]
it looks it also breaks the light of dawn quest.

**Comment by jackpoz** [comment]
This PR is more of its own separate project that will require fixes and tuning over the years, just like TC. 
We cannot expect it to be perfect but at the same time this means we cannot merge it.
It can stay here but I really think https://github.com/TrinityCore/TrinityCoreCustomChanges would be a better place (and I might ask to move it there sometime in the future)

**Comment by Lillecarl** [comment]
@kvipka how about making a pullrequest against trinitycorecustomchanges instead?

**Comment by Aokromes** [comment]
> Iirc there already is a branch for anti-cheat.

that's the anticheat i used to maintain, very diferent from this one.

**Comment by MaxtorCoder** [comment]
Ooohh, my bad then.

**Comment by warcrafte** [comment]
Bug Spell : [Master's Call] = If use, Pet Goes and you need again to use call pet spell.

Class Hunter

**Comment by peycho** [comment]
I would love to see this project revived and added to TC. Yes, I know is not perfect (yet), but it might be.

**Comment by illfated** [comment]
> I would love to see this project revived and added to TC. Yes, I know is not perfect (yet), but it might be.

Sure, but how would you feel if you find there are _lots_ of false positives and other bugs that makes it close to unusable for you, then finding out it takes several months to fix those bugs?

**Comment by Jinnaix** [comment]
To be honest, whoever runned a public server once knows that cheaters/hackers are a problem.
TrinityCore provides only basic warden defense that helps against stoneage stuff.
This system here might have flaws, maybe alot, but is a really good counter against the most popular cheats and hacks. 
You shouldn't use it on autokick/ban, but instead as a logger where you manually take actions afterwards. A flyhacker / teleporter / speed hacker doesn't only use his stuff for 2 minutes a day so you will be able to identify them.

**Comment by Aokromes** [comment]
> This system here might have flaws, maybe alot, but is a really good counter against the most popular cheats and hacks.
> You shouldn't use it on autokick/ban, but instead as a logger where you manually take actions afterwards. A flyhacker / teleporter / speed hacker doesn't only use his stuff for 2 minutes a day so you will be able to identify them.

The problem is when it breaks playability, ie flame leviathan stops working and so.

**Comment by Jinnaix** [comment]
Yes, these problems needs to be fixed of course

File: src/server/game/Entities/Player/Player.h [hunk 5]
```diff
@@ -1695,6 +1695,47 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         void BuildPlayerRepop();
         void RepopAtGraveyard();
 
+        /*********************************************************/
+        /***                  ANTICHEAT SYSTEM                   ***/
+        /*********************************************************///
+        // AntiCheat
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.h:1701]: It would be better to move all the anticheat stuff to its own class and call it from Player

**Comment by lastdarken** [comment]
Hi

error build

src/server/game/Accounts/AccountMgr.cpp:508:22: fatal error: cannot form pointer to deduced class template specialization type
    PreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
                     ^

**Comment by illfated** [comment]
@lastdarken : if this is something you only get from adding this PR content to your source, you should address your issue to the PR owner.

**Comment by Infernales** [comment]
@illfated I talked with the author of this PR. He has no motivation for fixing conflicts and compilation errors after new commits in TC, since this PR will never be merged with the 3.3.5 branch. This PR will only be updated if the author adds any new features. Thus, we take shovels and remove all the resulting shit on our own.
@lastdarken 
```
diff --git a/src/server/game/Accounts/AccountMgr.cpp b/src/server/game/Accounts/AccountMgr.cpp
index f9422a2..f60a6d5 100644
--- a/src/server/game/Accounts/AccountMgr.cpp
+++ b/src/server/game/Accounts/AccountMgr.cpp
@@ -505,7 +505,7 @@
 
 void AccountMgr::RecordAntiCheatLog(uint32 accountId, std::string const& name, std::string const& description, std::string const& position, uint32 realmid)
 {
-    PreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
+    LoginDatabasePreparedStatement* stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_ACCOUNT_ANTICHEAT);
     stmt->setUInt32(0, accountId);
     stmt->setString(1, name);
     stmt->setString(2, description);
diff --git a/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp b/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
index 7bd7a5b..94aa083 100644
--- a/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
+++ b/src/server/game/Movement/MovementGenerators/ChargeMovementGenerator.cpp
@@ -79,7 +79,7 @@
     owner->SetIsCharging(true);
     // Call for creature group update
     if (Creature* creature = owner->ToCreature())
-        creature->SignalFormationMovement(Position(_x, _y, _z), _movementId);
+        creature->SignalFormationMovement();
 }
 
 template<class T>
@@ -184,7 +184,7 @@
     owner->SetIsCharging(true);
     // Call for creature group update
     if (Creature* creature = owner->ToCreature())
-        creature->SignalFormationMovement(Position(_x, _y, _z), _movementId);
+        creature->SignalFormationMovement();
 }
 
 template<class T>
```
I did not deal with the conflict after the cfd63c3 commit because I don't have time for this.

**Comment by illfated** [comment]
OK, fair point. I see what you mean. Never mind,

**Comment by peycho** [comment]
@kvipka Thanks for the updates.

**Comment by lastdarken** [comment]
@kvipka  Thanks.

**Comment by jackpoz** [comment]
Added a link to https://github.com/kvipka/TrinityCore/tree/Antihacks at https://github.com/TrinityCore/TrinityCoreCustomChanges/wiki so users can find this branch.

Due to the reasons explained before, this PR cannot be merged to TC. Users are free to merge it locally and report issues to the author.

Thanks for the effort spent into it.

**Comment by Shadowfiende** [comment]
@kvipka

Please update to last version AC !


https://github.com/azerothcore/azerothcore-wotlk/pull/2088

**Comment by Shadowfiende** [comment]
Pro coder @kvipka  please updated for azerothcore

**Pull request closed without merging**